Delete empty origin directory when possible

szewai opened the pull request:
#### 9df9f747396e89b42e5d78d613fb09d7a067ae13
<pre>
Delete empty origin directory when possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=242118">https://bugs.webkit.org/show_bug.cgi?id=242118</a>

Reviewed by Chris Dumez.

We create origin directory and origin file when an origin is visited and it stores website data, but we do not delete
the directory and file when origin has no data any more (e.g. the origin clears its storage via storage API). To avoid
taking usre&apos;s disk space and putting pressure on backup server (e.g. rdar://94855540), let&apos;s delete origin file and
directory more eagerly. This patch tries to delete empty directories when OriginStorageManager (that mananges the
storage of an origin) is removed.

API tests: WebKit.DeleteEmptyOriginDirectoryWhenFetchData
           WebKit.DeleteEmptyOriginDirectoryWhenOriginIsGone

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::writeOriginToFile):
(WebKit::encode):
(WebKit::originDirectoryPath):
(WebKit::originFilePath):
(WebKit::isEmptyOriginDirectory):
(WebKit::deleteEmptyOriginDirectory):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::removeOriginStorageManagerIfPossible):
(WebKit::NetworkStorageManager::deleteDataOnDisk):
(WebKit::deleteOriginFileIfNecessary): Deleted.
(WebKit::NetworkStorageManager::deleteOriginDirectoryIfPossible): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::deleteEmptyDirectory):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
(WebKit::OriginStorageManager::path const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(runWebsiteDataStoreCustomPaths):
(TEST):

Canonical link: <a href="https://commits.webkit.org/252254@main">https://commits.webkit.org/252254@main</a>
</pre>

cdumez commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
```diff
@@ -51,6 +51,7 @@ class OriginStorageManager {
     void connectionClosed(IPC::Connection::UniqueID);
     bool persisted() const { return m_persisted; }
     void setPersisted(bool value);
+    String path() const { return m_path; }
```
Should return a `const String&`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
```diff
@@ -1082,3 +1082,77 @@ HTTPServer server([&] (Connection connection) {
+    receivedScriptMessage = false;
+    [webView loadHTMLString:htmlString baseURL:[NSURL URLWithString:@"https://apple.com"]];
+    TestWebKitAPI::Util::run(&receivedScriptMessage);
+    kill(pid, SIGKILL);
```
Is killing really the best way to achieve this? Could we simply navigate to another origin or about:blank?

szewai replied:
oh navigating to apple.com is actually enough (we are just checking for webkit.org, and loading apple.com here is doing what you say). I don't want to make the test rely on other features like process swap or process cache so I just kill the process to ensure the connection is lost.

cdumez reviewed (approved):
LGTM

szewai committed:
Delete empty origin directory when possible
https://bugs.webkit.org/show_bug.cgi?id=242118

Reviewed by Chris Dumez.

We create origin directory and origin file when an origin is visited and it stores website data, but we do not delete
the directory and file when origin has no data any more (e.g. the origin clears its storage via storage API). To avoid
taking usre's disk space and putting pressure on backup server (e.g. rdar://94855540), let's delete origin file and
directory more eagerly. This patch tries to delete empty directories when OriginStorageManager (that mananges the
storage of an origin) is removed.

API tests: WebKit.DeleteEmptyOriginDirectoryWhenFetchData
           WebKit.DeleteEmptyOriginDirectoryWhenOriginIsGone

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::writeOriginToFile):
(WebKit::encode):
(WebKit::originDirectoryPath):
(WebKit::originFilePath):
(WebKit::isEmptyOriginDirectory):
(WebKit::deleteEmptyOriginDirectory):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::removeOriginStorageManagerIfPossible):
(WebKit::NetworkStorageManager::deleteDataOnDisk):
(WebKit::deleteOriginFileIfNecessary): Deleted.
(WebKit::NetworkStorageManager::deleteOriginDirectoryIfPossible): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::deleteEmptyDirectory):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
(WebKit::OriginStorageManager::path const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(runWebsiteDataStoreCustomPaths):
(TEST):

Canonical link: https://commits.webkit.org/252254@main
